### PR TITLE
Visual mode (follow-on from #1413)

### PR DESCRIPTION
**Pull request by mrmr1993**
This uses the work done in #1413 to implement a bare-bones visual mode with hardcoded mappings. The implementation of the commands themselves is pulled directly from #1281.

Ping @smblott-github. All my changes are restricted to `content_scripts/mode_visual.coffee` and `content_scripts/visual_command.coffee`.

### CODE REVIEW

**Comment by aeschylus** [comment]
Assuming the tests pass, I dearly hope this can be merged. I have been waiting for this for a long time and am shamed not to have done it myself. Let me know if I can help.

**Comment by mrmr1993** [comment]
> I dearly hope this can be merged... Let me know if I can help.

There's quite a bit of work (to my mind at least) that needs to happen before #1413 can be merged. Since this is a direct follow on to that, anything you can do to help there would help speed the process along.

File: content_scripts/mode_visual.coffee [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        keyChar = KeyboardUtils.getKeyChar event
+        if (keyChar.match(/^[0-9]$/) and not event.shiftKey) or # Number prefixes.
+           (event.shiftKey and keyChar == "4") or # See #1417.
+           KeyboardUtils.getKeyChar(event) in ["h", "l", "k", "j", "b", "w", "0", "y"]
+          DomUtils.suppressPropagation event
+          @stopBubblingAndTrue
+        else
+          @suppressEvent
+
+      keypress: do ->
```
Comment (smblott-github) [inline content_scripts/mode_visual.coffee]: This might need to be `=>`.

(For `@exit`, see below.)

File: content_scripts/mode_visual.coffee [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+          else
+            numberPrefix ||= 1 # Make sure we don't do 0 repeats.
+            args = extendFocusArgs[keyChar]
+            command =
+              if args
+                "VisualCommand.extendFocus"
+              else if keyChar == "y"
+                args = []
+                "VisualCommand.yank"
+            Utils.invokeCommandString command, args for i in [0...numberPrefix]
```
Comment (smblott-github) [inline content_scripts/mode_visual.coffee]: On yank:

```coffeescript
@exit()
```

Also, probably better make sure that `numberPrefix` is `1`.

**Comment by smblott-github** [comment]
Looks like a nice start, @mrmr1993!

**Comment by mrmr1993** [comment]
Updated according to comments.

**Comment by mrmr1993** [comment]
This is stale. Closing.

**Pull request closed without merging**